### PR TITLE
Application Insights with Core Bot sample

### PR DESCRIPTION
**Pull request by munozemilio**
Enabled Application Insights in the Core Bot sample

### CODE REVIEW

File: samples/csharp_dotnetcore/csharp_dotnetcore.sln [hunk 1]
```diff
@@ -47,6 +47,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "StateMangementBot", "45.sta
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EchoBot", "02.echo-bot\EchoBot.csproj", "{586D7BC9-1A3B-4056-BD34-E83242524822}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CoreBot-App-Insights", "CoreBot-App-Insights\CoreBot-App-Insights.csproj", "{84584BDB-F423-45EB-96DD-C539A248E33F}"
```
Comment (daveta) [inline samples/csharp_dotnetcore/csharp_dotnetcore.sln]: >CoreBot-App-Insights [](start = 77, length = 20)

Not sure what the current trend, but are we still naming samples with number convention?

**Comment by sgellock** [comment]
@daveta @munozemilio yes, we are still using a numbering scheme for samples.  @vishwacsena can provide guidance if necessary

File: samples/csharp_dotnetcore/CoreBot-App-Insights/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+- [.NET Core SDK](https://dotnet.microsoft.com/download) version 2.1
+
+  ```bash
+  # determine dotnet version
+  dotnet --version
+  ```
+
+### Create a LUIS Application to enable language understanding
+
+LUIS language model setup, training, and application configuration steps can be found [here](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-howto-v4-luis?view=azure-bot-service-4.0&tabs=cs).
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/README.md]: >en-us [](start = 120, length = 5)

You can pull the "en-us" from all the links..

File: samples/csharp_dotnetcore/CoreBot-App-Insights/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+## To try this sample
+
+- Clone the repository
+
+    ```bash
+    git clone https://github.com/microsoft/botbuilder-samples.git
+    ```
+
+- In a terminal, navigate to `samples/csharp_dotnetcore/13.core-bot`
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/README.md]: >samples/csharp_dotnetcore/13.core-bo [](start = 30, length = 36)

According to Scott, we should get assigned a number..

File: samples/csharp_dotnetcore/CoreBot-App-Insights/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  ```bash
+  # run the bot
+  dotnet run
+  ```
+
+  B) Or from Visual Studio
+
+  - Launch Visual Studio
+  - File -> Open -> Project/Solution
+  - Navigate to `samples/csharp_dotnetcore/13.core-bot` folder
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/README.md]: >samples/csharp_dotnetcore/13.core-bo [](start = 17, length = 36)

Another one..

File: samples/csharp_dotnetcore/CoreBot-App-Insights/README.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+- Launch Bot Framework Emulator
+- File -> Open Bot
+- Enter a Bot URL of `http://localhost:3978/api/messages`
+
+## Deploy the bot to Azure
+
+To learn more about deploying a bot to Azure, see [Deploy your bot to Azure](https://aka.ms/azuredeployment) for a complete list of deployment instructions.
+
+## Further reading
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/README.md]: Might be nice to mention Application Insights

File: samples/csharp_dotnetcore/CoreBot-App-Insights/README.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# CoreBot
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/README.md]: Might be nice to reference Application Insights..

File: samples/csharp_dotnetcore/CoreBot-App-Insights/Dialogs/BookingDialog.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+{
+    public class BookingDialog : CancelAndHelpDialog
+    {
+        public BookingDialog()
+            : base(nameof(BookingDialog))
+        {
+            AddDialog(new TextPrompt(nameof(TextPrompt)));
+            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));
+            AddDialog(new DateResolverDialog());
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/Dialogs/BookingDialog.cs]: Add Telemetry?

File: samples/csharp_dotnetcore/CoreBot-App-Insights/Dialogs/MainDialog.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        public MainDialog(IConfiguration configuration, ILogger<MainDialog> logger, IBotTelemetryClient telemetryClient)
+            : base(nameof(MainDialog))
+        {
+            Configuration = configuration;
+            Logger = logger;
+            TelemetryClient = telemetryClient;
+
+            AddDialog(new TextPrompt(nameof(TextPrompt)));
+            AddDialog(new BookingDialog());
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/Dialogs/MainDialog.cs]: >WaterfallDialog [](start = 26, length = 15)

Add Telemetry?

File: samples/csharp_dotnetcore/CoreBot-App-Insights/Startup.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+            services.AddSingleton<ICredentialProvider, ConfigurationCredentialProvider>();
+
+            // Add Application Insights services into service collection
+            services.AddApplicationInsightsTelemetry();
+
+            // Create the telemetry client.
+            services.AddSingleton<IBotTelemetryClient, BotTelemetryClient>();
+
+            // Create http midleware
+            services.AddTransient<TelemetrySaveBodyASPMiddleware>();
```
Comment (daveta) [inline samples/csharp_dotnetcore/CoreBot-App-Insights/Startup.cs]: >TelemetrySaveBodyASPMiddleware [](start = 34, length = 30)

(Low Pri) If we're going to boot in this way, may make sense to add more docs to this middleware :)  This was wrapped in the other helper before.  But like you mention, generally useful so makes sense to expose..

**Comment by daveta** [comment]
Great job @munozemilio !!

**Comment by cleemullins** [comment]
Fixed (part of) #1480

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/CancelAndHelpDialog.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            var result = await InterruptAsync(innerDc, cancellationToken);
+            if (result != null)
+            {
+                return result;
+            }
+
+            return await base.OnContinueDialogAsync(innerDc, cancellationToken);
+        }
+
+        private async Task<DialogTurnResult> InterruptAsync(DialogContext innerDc, CancellationToken cancellationToken)
```
Comment (scheyal) [inline samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/CancelAndHelpDialog.cs]: Need to instrument help & cancel for application insights telemetry.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+using Microsoft.Recognizers.Text.DataTypes.TimexExpression;
+
+namespace Microsoft.BotBuilderSamples.Dialogs
+{
+    public class DateResolverDialog : CancelAndHelpDialog
+    {
+        public DateResolverDialog(string id = null)
+            : base(id ?? nameof(DateResolverDialog))
+        {
+            AddDialog(new DateTimePrompt(nameof(DateTimePrompt), DateTimePromptValidator));
```
Comment (scheyal) [inline samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs]: Are these produce application insights telemetry? if not, we need to add telemetry here as well.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/CancelAndHelpDialog.cs (hunk 10 above)
Comment (munozemilio) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/CancelAndHelpDialog.cs]: Cancel&Help Dialog is a Base class, the one extending is DateResolver and BookingDialog and both are set with telemetryclient

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs (hunk 11 above)
Comment (munozemilio) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs]: The telemetry client is set when added to the DialogSet

**Pull request merged**